---
title: "Non-overlapping Windows Analysis"
---

### Initialization
```{r initialization}
if (length(params$window_size) == 0) {
  print("Window size is not specified. Quitting...")
  knitr::knit_exit()
}

fn_alisim <- paste(getwd(), "/simulation/", params$prefix, ".fa", sep="")
window_outdir <- paste(getwd(), "/windows/", sep="")

if (!file.exists(window_outdir)) {
  dir.create(window_outdir, recursive = T)
}

# read fasta file
s <- read.fasta(fn_alisim, whole.header = T)
lens <- getLength(s)[1]
options(scipen=999)

print("Generating window alignments...")
for (w in params$window_size) {
  if (w == 0) {
    print("Window size cannot be zero. Moving on...")
    next
  }
  
  temp_outdir <- paste(window_outdir,w,"/alignment/",sep="")
  if (!file.exists(temp_outdir)) {
    dir.create(temp_outdir, recursive = T)
  }
  
  if (sapply(temp_outdir,function(dir){length(list.files(dir,pattern='fa'))}) != 0 && params$redo == FALSE) {
    print(paste(w, "bp window alignment already exists. Moving on...", sep=""))
    next
  } 
  
  start <- 1
  for (i in 1:floor(lens/w)) {
    subfasta <- lapply(s, function(x) x[seq(from = start, to = as.numeric(i*w))])
    df <- do.call(rbind,subfasta)
      
    wi <- floor(log(floor(lens/w)) / log(10)) + 1
    li <- formatC(i, width = wi, format = "d", flag = "0")
      
    subfasta <- setNames(split(df, seq(nrow(df))), rownames(df))
    write.fasta(sequences=subfasta, names=names(subfasta), file.out=paste(temp_outdir,"window_",li,".fa",sep=""), nbchar = 100)
      
    start <- start + w
  }
}
```

```{r windowtree}
window_outdir <- paste(getwd(), "/windows/", sep="")
window_sizes <- list.dirs(window_outdir, full.names = F, recursive = F)

print("Generating window trees...")
for (i in window_sizes) {
  temp_outdir <- paste(window_outdir,i,"/trees/",sep="")
  if (!file.exists(temp_outdir)) {
    dir.create(temp_outdir, recursive = T)
  }
  
  if (sapply(temp_outdir,function(dir){length(list.files(dir,pattern='treefile'))}) != 0 && params$redo == FALSE) {
    print(paste(i, "bp window treefile already exists. Moving on...", sep=""))
    next
  }
  
  fastadir <- paste(window_outdir,i,"/alignment/",sep="")
  prefix <- paste(temp_outdir,params$prefix,sep="")

  if (is.null(params$outgroup) || params$outgroup == ""){
    system(paste(params$iqtree2dir,"-S",fastadir,"-m",params$alisim_model,"-pre",prefix,"--quiet", sep=" "))
  } else {
    system(paste(params$iqtree2dir,"-S",fastadir,"-m",params$alisim_model,"-pre",prefix,"--quiet -o",params$outgroup, sep=" "))
  }
}

```

### Annotation
```{r annotation}
window_outdir <- paste(getwd(), "/windows/", sep="")
window_sizes <- list.dirs(window_outdir, full.names = F, recursive = F)

fn_uqtops <- paste(getwd(), "/simulation/", params$prefix, ".uqtops", sep="")
options(scipen=999)

print("Reading unique topologies...")
t_hash <- hash()
t_count <- 1

topology_file <- file(description=fn_uqtops, open="r", blocking = TRUE)
repeat {
  tl <- readLines(topology_file, n=1, warn=F)
  if (identical(tl, character(0))) {
    close(topology_file)
    break
  }
  
  tl <- write.tree(unroot(read.tree(text=tl)))
  t_hash[tl] = paste("T",t_count,sep="")
  t_count <- t_count+1
}

print("Annotating non-overlapping windows...")
for (i in window_sizes) {
  temp_outdir <- paste(window_outdir,i,"/summary/",sep="")
  if (!file.exists(temp_outdir)) {
    dir.create(temp_outdir, recursive = T)
  }
  
  fn_summary <- paste(temp_outdir, params$prefix, ".atsum", sep="")
  fn_count <- paste(temp_outdir, params$prefix, ".count", sep="")
  fn_figure <- paste(temp_outdir, params$prefix, ".atsum.tiff", sep="")
  fn_cfigure <- paste(temp_outdir, params$prefix, ".count.tiff", sep="")
  fn_tops <- paste(temp_outdir, params$prefix, ".topsum", sep="")
  if (all(file.exists(fn_summary, fn_figure, fn_count, fn_cfigure, fn_tops)) && params$redo == FALSE) {
    print(paste(i, "bp window annotation already exists. Moving on...", sep=""))
    next
  }
  
  # reading each fasta file and its respective topology
  seq_list <- sort(list.files(paste(window_outdir, i, "/alignment/", sep="")))
  iqtree_file <- file(description = paste(window_outdir, i, "/trees/", params$prefix, ".treefile", sep=""), open="r", blocking = TRUE)
  
  seq_len <- data.frame()
  iterator <- 1
  start <- 0
  
  repeat {
    # set the topology based on topology map
    ttop <- "NT"
    trl <- readLines(iqtree_file, n=1)
    if (identical(trl, character(0))) {
      colnames(seq_len) <- c("source","start","stop","length","newick","topology")
      close(iqtree_file)
      break
    }
    
    tl <- read.tree(text=trl)
    tl$edge.length <- NULL
    tree <- write.tree(tl)
    
    for (j in ls(t_hash)) {
      if (all.equal.phylo(read.tree(text = j), tl)) {
        ttop <- as.character(values(t_hash[j]))
      }
    }
    
    # read the sequence file
    pl <- seq_list[iterator]
    seq_len <- rbind(seq_len, c(pl,start+1,start+as.numeric(i),as.numeric(i),tree, ttop))
    
    start <- start+as.numeric(i)
    iterator <- iterator+1
  }
  
  write.table(seq_len, fn_summary, quote=F, sep="\t", row.names=F, col.names=T)
  
  # extract top topologies
  top_tops <- seq_len %>% 
    group_by(newick) %>% 
    summarise(count = n()) %>% 
    arrange(desc(count)) %>%
    mutate(cum.percentage = round(cumsum(count)/nrow(seq_len),3))
  
  write.table(top_tops, fn_tops, quote=F, sep="\t", row.names=F, col.names=T)
  
  # count contiguous blocks per topology
  count_contiguous <- seq_len %>% 
    group_by(topology,
             group_run = data.table::rleid(topology)) %>% 
    summarise(count = n()) %>% 
    arrange(group_run)
  count_contiguous$group_run <- NULL
  
  write.table(count_contiguous, fn_count, quote=F, row.names=F)
  
  cons_windows <- count_contiguous %>%
    group_by(topology, count) %>%
    summarise(total = n())
  
  tiff(filename=fn_cfigure, units="px", width=2880, height=1800)
  print(ggplot(cons_windows, aes(fill=topology)) + 
        geom_bar(position="dodge", stat="identity", aes(y=total, x=count)) +
        aes(x = fct_inorder(topology)) +
        ggtitle("Distribution of Consecutive Windows Per Topology") + 
        facet_grid(~topology) +
        theme(
          plot.title = element_text(hjust = 0.5, size = 50),
          axis.title.x=element_blank(),
          axis.title.y=element_blank(),
          axis.text.y=element_text(size=30),
          axis.text.x=element_text(size=30),
          legend.title=element_blank(),
          legend.text=element_text(size=30),
          legend.key.size=unit(2,"cm")
        ))
  dev.off()

  # extracting essential columns
  dat <- apply(seq_len, 1,
               function(x) data.table(bp = x["start"]:x["stop"], topology = x["topology"])) %>%
    rbindlist()

  # plotting the chromosome using random color
  print(paste("Plotting for",i,"bp windows..."))
  tiff(file=fn_figure, units="px", width=250, height=1800)
  print(ggplot() +
    geom_rect(data=dat,aes(fill=topology,ymin=bp-1,ymax=bp,xmin=0,xmax=1)) +
    labs(x=paste(i,"bp")) +
    scale_fill_brewer(palette="Set1") +
    theme(panel.background = element_rect(fill = "white", colour = "white", size = 0.5, linetype = "solid")) +
    theme(
      axis.title.x=element_blank(),
      axis.text.x=element_blank(),
      axis.ticks.x=element_blank(),
      axis.text.y=element_text(size=20),
      legend.title=element_blank(),
      legend.text=element_text(size=20),
      legend.key.size=unit(2,"cm")
      ))
  dev.off()
}
```

```{r contiguous}
window_outdir <- paste(getwd(), "/windows/", sep="")
window_sizes <- list.dirs(window_outdir, full.names = F, recursive = F)

fn_alisim <- paste(getwd(), "/simulation/", params$prefix, ".fa", sep="")

# read fasta file
s <- read.fasta(fn_alisim, whole.header = T)
sdf <- do.call(rbind,s)

# combine windows based on topology
for (i in window_sizes) {
  temp_outdir <- paste(window_outdir,i,"/contiguous/alignment/",sep="")
  if (!file.exists(temp_outdir)) {
    dir.create(temp_outdir, recursive = T)
  }
  
  treedir <- paste(window_outdir,i,"/contiguous/trees/",sep="")
  if (file.exists(treedir)==F) {
    dir.create(treedir, recursive = T)
  }
  
  acount <- sapply(temp_outdir,function(dir){length(list.files(dir,pattern='fa'))})
  tcount <- sapply(treedir,function(dir){length(list.files(dir,pattern='iqtree'))})
  
  if (acount != 0 && tcount != 0 && params$redo == FALSE) {
    print(paste(i, "bp window alignment and tree already exist. Moving on...", sep=""))
    next
  }
  
  fn_summary <- paste(window_outdir,i,"/summary/",params$prefix,".atsum", sep="")
  
  smfile <- read.table(fn_summary)
  colnames(smfile) <- smfile[1,]
  smfile <- smfile[-1,]
    
  unique_topology <- sort(unique(smfile$newick))
  newick_map <- data.frame()
  count <- 1
  
  print("Generating topology-based contiguous sequence...")
  for (t in unique_topology) {
    sub_sw <- subset(smfile[smfile$newick == t,])
    align_idx <- unlist(Map(seq, sub_sw$start, sub_sw$stop))
        
    alignment <- sdf[,align_idx]
    subfasta <- setNames(split(alignment, seq(nrow(alignment))), rownames(alignment))
      
    write.fasta(sequences=subfasta, names=names(subfasta), file.out=paste(temp_outdir,"G",count,".fa",sep=""), nbchar = 100)
    newick_map <- rbind(newick_map, c(t, paste("G",count,sep="")))
    count <- count + 1
  }
    
  write.table(newick_map, paste(temp_outdir, "../", params$prefix, ".map", sep=""), quote = F, row.names = F, col.names = F)
    
  print("Generating contiguous window trees...")
  prefix <- paste(treedir,params$prefix, sep="")
  if (is.null(params$outgroup) || params$outgroup == ""){
    system(paste(params$iqtree2dir,"-S",temp_outdir,"-m",params$alisim_model,"-pre",prefix,"--quiet -T 2", sep=" "))
  } else {
    system(paste(params$iqtree2dir,"-S",temp_outdir,"-m",params$alisim_model,"-pre",prefix,"--quiet -T 2 -o",params$outgroup, sep=" "))
  }
}
```