---
title: "Stepwise NOW for Empirical Dataset"
---

# Summary of stepwise non-overlapping windows
```{r summary-setup, include=FALSE}
write.table(c("",
              "####################################",
              "####           Summary          ####",
              "####################################"),
            file=fn_log, quote=F, row.names=F, col.names=F, append=T)
```

```{r delta-aic, include=FALSE}
write.table(c("",
              "--------- Delta AIC per Step -------"),
              file=fn_log, quote=F, row.names=F, col.names=F, append=T)

# create doSNOW cluster
nwcl <- makeCluster(stepthread)
doSNOW::registerDoSNOW(nwcl)

# window alignments for shorter window size
foreach (i = 1:1) %dopar% {
  options(scipen=999)

  # output file
  df_output <- data.table::data.table()
  fn_output <- paste0(getwd(), "/", i, "/", i, ".aic.sum")

  # input directory
  dir_long_filtered <- paste0(getwd(), "/", i, "/", wsizes[i], "/filtered/")
  dir_short_filtered <- paste0(getwd(), "/", i, "/", wsizes[i+1], "/filtered/")
  dir_short_perwindow <- paste0(getwd(), "/", i, "/", wsizes[i+1], "/perwindow/")

  # extract the number of windows
  long_numw <- s_ln_new/wsizes[i]
  short_numw <- s_ln_new/wsizes[i+1]
  long_width <- ceiling(log(long_numw) / log(10)) + 1
  short_width <- ceiling(log(short_numw) / log(10)) + 1

  # iterate over long windows
  start <- 1

  for (j in 1:long_numw) {
    # name of windows
    long_name <- paste0("window_", formatC(j, width=long_width, format="d", flag="0"))
    short_name_left <- paste0("window_", formatC((j*2)-1, width=short_width, format="d", flag="0"))
    short_name_right <- paste0("window_", formatC(j*2, width=short_width, format="d", flag="0"))

    # iqtree file for long window
    fn_long_iqtree <- paste0(dir_long_filtered, long_name, "/", long_name, ".fa.iqtree")
    
    if (!file.exists(fn_long_iqtree)) {
      if (i == 1) {
        fn_long_iqtree <- paste0(getwd(), "/", i, "/", wsizes[i], "/perwindow/", long_name, "/", long_name, ".fa.iqtree")
      } else {
        fn_long_iqtree <- paste0(getwd(), "/", i-1, "/", wsizes[i], "/perwindow/", long_name, "/", long_name, ".fa.iqtree")
      }
    }

    # iqtree files for short windows
    fn_short_left_iqtree <- paste0(dir_short_filtered, short_name_left, "/", short_name_left, ".fa.iqtree")
    fn_short_right_iqtree <- paste0(dir_short_filtered, short_name_right, "/", short_name_right, ".fa.iqtree")

    if (!file.exists(fn_short_left_iqtree)) {
      fn_short_left_iqtree <- paste0(dir_short_perwindow, short_name_left, "/", short_name_left, ".fa.iqtree")
    }

    if (!file.exists(fn_short_right_iqtree)) {
      fn_short_right_iqtree <- paste0(dir_short_perwindow, short_name_right, "/", short_name_right, ".fa.iqtree")
    }

    # check if iqtree file exists
    if (!file.exists(fn_long_iqtree)) {
        # update output data.table
        df_output <- rbind(df_output,
                           list(long=long_name, short_left=short_name_left, short_right=short_name_right,
                                start=start, stop=j*wsizes[i],
                                long_aic="NA", short_aic="NA", delta_aic="NA"))
    } else {
        # get the aic for all windows
        long_aic <- f_calculate_aic_from_iqtree(fn_long_iqtree)
        short_aic <- f_calculate_aic_from_iqtree(c(fn_short_left_iqtree, fn_short_right_iqtree))

        # update output data.table
        df_output <- rbind(df_output,
                           list(long=long_name, short_left=short_name_left, short_right=short_name_right,
                                start=start, stop=j*wsizes[i],
                                long_aic=long_aic, short_aic=short_aic, delta_aic=short_aic-long_aic))
    }

    start <- start + wsizes[i]
  }

  # save output file
  data.table::fwrite(df_output, file=fn_output, quote=FALSE, sep="\t")
  log4r::info(fn_logger, paste0("Per-window AIC summary added: Step ", i, "."))
}

stopCluster(nwcl)
```

## Window-based delta AIC across chromosome {.tabset}
```{r, results='asis', out.width="100%", echo=FALSE}
for (i in 1:1) {
    fn_output <- paste0(getwd(), "/", i, "/", i, ".aic.sum")
    if (!file.exists(fn_output)) {
        log4r::warn(fn_logger, paste0("File not found: per-window AIC summary for step ", i, ". Skipped."))
        next
    }

    # open AIC summary file
    df_aic_sum <- data.table::fread(fn_output)
    
    # visualization
    cat(paste('###', wsizes[i], 'vs.', wsizes[i+1], '\n'))
    plot(f_chromosomal_delta_aic(df_aic_sum, wsizes[i], wsizes[i+1]))
    cat('  \n\n')
}
```

## Window-based delta AIC for selected windows {.tabset}
```{r, results='asis', out.width="100%", echo=FALSE}
options(scipen=999)

for (i in 1:1) {
    fn_output <- paste0(getwd(), "/", i, "/", i, ".aic.sum")
    if (!file.exists(fn_output)) {
        log4r::warn(fn_logger, paste0("File not found: per-window AIC summary for step ", i, ". Skipped."))
        next
    }

    # input directory
    dir_long_filtered <- paste0(getwd(), "/", i, "/", wsizes[i], "/filtered/")
    dir_short_filtered <- paste0(getwd(), "/", i, "/", wsizes[i+1], "/filtered/")

    # open AIC summary file
    df_aic_sum <- data.table::fread(fn_output)
    df_aic_subset <- df_aic_sum %>%
        filter(delta_aic != "NA") %>%
        arrange(delta_aic)
    
    # extract random indices
    indices <- floor(seq(1, nrow(df_aic_subset), length.out = 2))

    # visualization
    cat(paste('###', wsizes[i], 'vs.', wsizes[i+1], '{.tabset .tabset-pills} \n'))
    
    iterator <- 1
    for (j in indices) {
        cat(paste('####', iterator, '\n'))
        cat(paste0('Long: ', df_aic_subset$long[j], '  \n'))
        cat(paste0('Short: ', df_aic_subset$short_left[j], ', ', df_aic_subset$short_right[j], '  \n'))
        cat(paste0('Position: ', df_aic_subset$start[j], '-', df_aic_subset$stop[j], 'bp  \n'))
        cat(paste0('Delta AIC: ', df_aic_subset$delta_aic[j], '  \n\n'))

        # iqtree files
        fn_long_iqtree <- paste0(dir_long_filtered, df_aic_subset$long[j], "/", df_aic_subset$long[j], ".fa.treefile")
        fn_short_left_iqtree <- paste0(dir_short_filtered, df_aic_subset$short_left[j], "/", df_aic_subset$short_left[j], ".fa.treefile")
        fn_short_right_iqtree <- paste0(dir_short_filtered, df_aic_subset$short_right[j], "/", df_aic_subset$short_right[j], ".fa.treefile")

        # plotting
        ls_treefile <- c(fn_long_iqtree, fn_short_left_iqtree, fn_short_right_iqtree)
        ls_annotation <- c(paste0(wsizes[i],"bp"), paste0(wsizes[i+1],"bp (Left)"), paste0(wsizes[i+1],"bp (Right)"))
        plot(f_plot_multiple_trees(ls_treefile, ls_annotation, min_branch_support))
        cat('  \n\n')

        # fasta files
        fn_short_left_fa <- paste0(dir_short_filtered, df_aic_subset$short_left[j], "/", df_aic_subset$short_left[j], ".fa")
        fn_short_right_fa <- paste0(dir_short_filtered, df_aic_subset$short_right[j], "/", df_aic_subset$short_right[j], ".fa")

        # print alignments
        cat(paste0('##### ', wsizes[i+1], "bp (Left)", '  \n'))
        f_print_fasta_alignment(fn_short_left_fa)
        cat(paste0('##### ', wsizes[i+1], "bp (Right)", '  \n'))
        f_print_fasta_alignment(fn_short_right_fa)
        cat('  \n\n')

        # update variables
        iterator <- iterator + 1
    }

    cat('  \n\n')
}
```